test/hotspot/jtreg/serviceability/jvmti/vthread/ToggleNotifyJvmtiTest/ToggleNotifyJvmtiTest.java

```diff
@@ -0,0 +1,206 @@
+ */
+public class ToggleNotifyJvmtiTest {
+    private static final int VTHREADS_CNT = 20;
+    private static final String AGENT_LIB = "ToggleNotifyJvmtiTest";
+    private static final WhiteBox WB = WhiteBox.getWhiteBox();
+
+    private static native boolean IsAgentStarted();
+    private static native int VirtualThreadStartedCount();
+    private static native int VirtualThreadEndedCount();
+    private static native int ThreadStartedCount();
+    private static native int ThreadEndedCount();
+
+    static void log(String str) { System.out.println(str); }
+
+    static public void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Interruption in TestedThread.sleep: \n\t" + e);
+        }
+    }
+
+    static TestedThread[] threads = new TestedThread[VTHREADS_CNT];
+    static Thread vts[] = new Thread[VTHREADS_CNT];
+
+    static private void startThread(int i) {
+        String name = "TestedThread" + i;
+        TestedThread thread = new TestedThread(name);
+        vts[i] = Thread.ofVirtual().name(name).start(thread);
+        thread.ensureReady();
+        threads[i] = thread;
+        log("# Java: started vthread: " + name);
+    }
+
+    static synchronized private void startThreads() {
+        log("\n# Java: Starting vthreads");
+        for (int i = 0; i < VTHREADS_CNT; i++) {
+            sleep(1);
+            startThread(i);
+        }
+    }
+
+    static private synchronized void finishThreads() {
+        try {
+            for (int i = 0; i < VTHREADS_CNT; i++) {
+                TestedThread thread = threads[i];
+                thread.letFinish();
+                vts[i].join();
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static private void setVirtualThreadsNotifyJvmtiMode(int iter, boolean enable) {
+        boolean status = WB.setVirtualThreadsNotifyJvmtiMode(enable);
+        if (!status) {
+            throw new RuntimeException("Java: failed to set VirtualThreadsNotifyJvmtiMode: " + enable);
+        }
+        log("# main: SetNotifyJvmtiEvents: #" + iter + " enable: " + enable);
+    }
+
+    // Accumulative results after each finished test cycle.
+    static private void printResults() {
+        log("  VirtualThreadStart events: " + VirtualThreadStartedCount());
+        log("  VirtualThreadEnd events:   " + VirtualThreadEndedCount());
+        log("  ThreadStart events:        " + ThreadStartedCount());
+        log("  ThreadEnd events:          " + ThreadEndedCount());
+    }
+
+    static private void run_test_cycle(int iter) throws Exception {
+        log("\n# Java: Started test cycle #" + iter);
+
+        // Disable notifyJvmti events mode at test cycle start.
+        // It is unsafe to do so if any virtual threads are executed.
+        setVirtualThreadsNotifyJvmtiMode(iter, false);
```

Comment:
What happens if the main thread is a virtual thread because the virtual thread wrapper is being used? Is there any concern with disabling notifications in this case?

Reply:
I've tested it and it is kind of working now.
But I'd like to minimize risk now and test something which is not really needed.
So, this is why the testing cycles are still there.
The `serviceability/jvmti/vthread` tests are not supposed to be run with the virtual thread wrapper.
It is why we have the special problem list `ProblemList-vthread.txt` in Loom repository.
I guess, we should replicate the same or alike approach in the main line when virtual thread wrapper is introduced.